File: Packages/HistoryKit/Tests/HistoryKitTests/HistoryStoreTests.swift
import TestKit
import Foundation
@testable import HistoryKit

private func makeTempDir() -> URL {
    FileManager.default.temporaryDirectory
        .appendingPathComponent("HistoryStoreTests-\(UUID().uuidString)", isDirectory: true)
}

let historyStoreTests: [TestCase] = [
    TestCase("addScreenshotWritesCopyThumbAndIndex") { t in
        let dir = makeTempDir()
        defer { try? FileManager.default.removeItem(at: dir) }
        let store = HistoryStore(directory: dir, cap: 50, maxAge: 30 * 24 * 60 * 60)
        guard let entry = t.unwrap(store.addScreenshot(pngData: makePNGData(), cap: 50, maxAge: 30 * 24 * 60 * 60)) else { return }
        t.equal(entry.kind, .screenshot)
        t.notNil(entry.imageFile)
        t.isNil(entry.filePath)
        t.isTrue(FileManager.default.fileExists(atPath: store.thumbURL(for: entry).path))
        guard let imageURL = t.unwrap(store.imageURL(for: entry)) else { return }
        t.isTrue(FileManager.default.fileExists(atPath: imageURL.path))
        t.isTrue(FileManager.default.fileExists(
            atPath: dir.appendingPathComponent("history.json").path))
        t.equal(store.index.entries.count, 1)
    },
    TestCase("addRecordingStoresReferenceNotCopy") { t in
        let dir = makeTempDir()
        defer { try? FileManager.default.removeItem(at: dir) }
        let saved = FileManager.default.temporaryDirectory
            .appendingPathComponent("rec-\(UUID().uuidString).mp4")
        FileManager.default.createFile(atPath: saved.path, contents: Data([0x0]))
        defer { try? FileManager.default.removeItem(at: saved) }
        let store = HistoryStore(directory: dir, cap: 50, maxAge: 30 * 24 * 60 * 60)
        guard let entry = t.unwrap(store.addRecording(
            filePath: saved.path, thumbnailSource: makePNGData(), cap: 50, maxAge: 30 * 24 * 60 * 60)) else { return }
        t.equal(entry.kind, .recording)
        t.isNil(entry.imageFile)
        t.equal(entry.filePath, saved.path)
        t.isTrue(FileManager.default.fileExists(atPath: store.thumbURL(for: entry).path))
    },
    TestCase("reloadRoundTripsIndex") { t in
        let dir = makeTempDir()
        defer { try? FileManager.default.removeItem(at: dir) }
        let store = HistoryStore(directory: dir, cap: 50, maxAge: 30 * 24 * 60 * 60)
        store.addScreenshot(pngData: makePNGData(), cap: 50, maxAge: 30 * 24 * 60 * 60)
        store.addScreenshot(pngData: makePNGData(), cap: 50, maxAge: 30 * 24 * 60 * 60)
        let reloaded = HistoryStore(directory: dir, cap: 50, maxAge: 30 * 24 * 60 * 60)
        t.equal(reloaded.index.entries.map(\.id), store.index.entries.map(\.id))
    },
    TestCase("capEvictionDeletesOwnedFiles") { t in
        let dir = makeTempDir()
        defer { try? FileManager.default.removeItem(at: dir) }
        let store = HistoryStore(directory: dir, cap: 1, maxAge: 30 * 24 * 60 * 60)
        guard let first = t.unwrap(store.addScreenshot(pngData: makePNGData(), cap: 1, maxAge: 30 * 24 * 60 * 60)) else { return }
        let firstImage = store.imageURL(for: first)!
        let firstThumb = store.thumbURL(for: first)
        store.addScreenshot(pngData: makePNGData(), cap: 1, maxAge: 30 * 24 * 60 * 60)
        t.equal(store.index.entries.count, 1)
        t.isFalse(FileManager.default.fileExists(atPath: firstImage.path))
        t.isFalse(FileManager.default.fileExists(atPath: firstThumb.path))
    },
    TestCase("removeNeverDeletesSavedRecordingFile") { t in
        let dir = makeTempDir()
        defer { try? FileManager.default.removeItem(at: dir) }
        let saved = FileManager.default.temporaryDirectory
            .appendingPathComponent("rec-\(UUID().uuidString).mp4")
        FileManager.default.createFile(atPath: saved.path, contents: Data([0x0]))
        defer { try? FileManager.default.removeItem(at: saved) }
        let store = HistoryStore(directory: dir, cap: 50, maxAge: 30 * 24 * 60 * 60)
        guard let entry = t.unwrap(store.addRecording(
            filePath: saved.path, thumbnailSource: makePNGData(), cap: 50, maxAge: 30 * 24 * 60 * 60)) else { return }
        let thumb = store.thumbURL(for: entry)
        store.remove(id: entry.id)
        t.isTrue(store.index.entries.isEmpty)
        t.isFalse(FileManager.default.fileExists(atPath: thumb.path))
        t.isTrue(FileManager.default.fileExists(atPath: saved.path),
                 "user's saved recording must survive history delete")
    },
    TestCase("clearAllEmptiesIndexAndDeletesOwnedFiles") { t in
        let dir = makeTempDir()
        defer { try? FileManager.default.removeItem(at: dir) }
        let store = HistoryStore(directory: dir, cap: 50, maxAge: 30 * 24 * 60 * 60)
        guard let e = t.unwrap(store.addScreenshot(pngData: makePNGData(), cap: 50, maxAge: 30 * 24 * 60 * 60)) else { return }
        let image = store.imageURL(for: e)!
        store.clearAll()
        t.isTrue(store.index.entries.isEmpty)
        t.isFalse(FileManager.default.fileExists(atPath: image.path))
        let reloaded = HistoryStore(directory: dir, cap: 50, maxAge: 30 * 24 * 60 * 60)
        t.isTrue(reloaded.index.entries.isEmpty)
    },
    TestCase("corruptIndexStartsEmpty") { t in
        let dir = makeTempDir()
        defer { try? FileManager.default.removeItem(at: dir) }
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        try? Data("garbage".utf8).write(to: dir.appendingPathComponent("history.json"))
        let store = HistoryStore(directory: dir, cap: 50, maxAge: 30 * 24 * 60 * 60)
        t.isTrue(store.index.entries.isEmpty)
    },
    TestCase("missingRecordingFilePrunedAtLoad") { t in
        let dir = makeTempDir()
        defer { try? FileManager.default.removeItem(at: dir) }
        let saved = FileManager.default.temporaryDirectory
            .appendingPathComponent("rec-\(UUID().uuidString).mp4")
        FileManager.default.createFile(atPath: saved.path, contents: Data([0x0]))
        let store = HistoryStore(directory: dir, cap: 50, maxAge: 30 * 24 * 60 * 60)
        store.addRecording(filePath: saved.path, thumbnailSource: makePNGData(), cap: 50, maxAge: 30 * 24 * 60 * 60)
        try? FileManager.default.removeItem(at: saved)   // user deletes it in Finder
        let reloaded = HistoryStore(directory: dir, cap: 50, maxAge: 30 * 24 * 60 * 60)
        t.isTrue(reloaded.index.entries.isEmpty)
    },
    TestCase("agePruneAppliesAtLoad") { t in
        let dir = makeTempDir()
        defer { try? FileManager.default.removeItem(at: dir) }
        let store = HistoryStore(directory: dir, cap: 50, maxAge: 30 * 24 * 60 * 60)
        store.addScreenshot(pngData: makePNGData(), cap: 50, maxAge: 30 * 24 * 60 * 60,
                            date: Date().addingTimeInterval(-31 * 86_400))
        t.equal(store.index.entries.count, 1)
        let reloaded = HistoryStore(directory: dir, cap: 50, maxAge: 30 * 24 * 60 * 60)
        t.isTrue(reloaded.index.entries.isEmpty)
    },
    TestCase("neverRetentionKeepsOldEntriesAtLoad") { t in
        let dir = makeTempDir()
        defer { try? FileManager.default.removeItem(at: dir) }
        let store = HistoryStore(directory: dir, cap: 50, maxAge: nil)
        store.addScreenshot(pngData: makePNGData(), cap: 50, maxAge: nil,
                            date: Date().addingTimeInterval(-365 * 86_400))
        let reloaded = HistoryStore(directory: dir, cap: 50, maxAge: nil)
        t.equal(reloaded.index.entries.count, 1)
    },
    TestCase("pruneExpiredDropsEntriesPastRetentionAndDeletesCachedFiles") { t in
        let dir = makeTempDir()
        defer { try? FileManager.default.removeItem(at: dir) }
        // Seed with retention off, so the second add can't evict the first at
        // insert time — this test is about the sweep, not the on-add prune.
        let store = HistoryStore(directory: dir, cap: 50, maxAge: nil)
        let now = Date()
        guard let old = t.unwrap(store.addScreenshot(pngData: makePNGData(), cap: 50,
                                                     maxAge: nil,
                                                     date: now.addingTimeInterval(-1801))),
              let fresh = t.unwrap(store.addScreenshot(pngData: makePNGData(), cap: 50,
                                                       maxAge: nil, date: now))
        else { return }
        let oldImage = dir.appendingPathComponent(old.imageFile ?? "")
        t.isTrue(FileManager.default.fileExists(atPath: oldImage.path))

        t.isTrue(store.pruneExpired(cap: 50, maxAge: 1800, now: now))
        t.equal(store.index.entries.map(\.id), [fresh.id])
        // the expired capture's cached copy + thumbnail are gone from disk
        t.isFalse(FileManager.default.fileExists(atPath: oldImage.path))
        t.isFalse(FileManager.default.fileExists(
            atPath: dir.appendingPathComponent(old.thumbFile).path))
        // nothing left to evict -> no republish
        t.isFalse(store.pruneExpired(cap: 50, maxAge: 1800, now: now))
    },
    TestCase("pruneExpiredNeverTouchesARecordingsSavedFile") { t in
        let dir = makeTempDir()
        defer { try? FileManager.default.removeItem(at: dir) }
        let saved = FileManager.default.temporaryDirectory
            .appendingPathComponent("rec-\(UUID().uuidString).mp4")
        FileManager.default.createFile(atPath: saved.path, contents: Data([0x0]))
        defer { try? FileManager.default.removeItem(at: saved) }
        let store = HistoryStore(directory: dir, cap: 50, maxAge: 1800)
        let now = Date()
        store.addRecording(filePath: saved.path, thumbnailSource: makePNGData(), cap: 50,
                           maxAge: 1800, date: now.addingTimeInterval(-1801))
        t.isTrue(store.pruneExpired(cap: 50, maxAge: 1800, now: now))
        t.isTrue(store.index.entries.isEmpty)
        t.isTrue(FileManager.default.fileExists(atPath: saved.path))
    },
    TestCase("savedFileExistsReflectsDisk") { t in
        let dir = makeTempDir()
        defer { try? FileManager.default.removeItem(at: dir) }
        let saved = FileManager.default.temporaryDirectory
            .appendingPathComponent("rec-\(UUID().uuidString).mp4")
        FileManager.default.createFile(atPath: saved.path, contents: Data([0x0]))
        let store = HistoryStore(directory: dir, cap: 50, maxAge: 30 * 24 * 60 * 60)
        guard let entry = t.unwrap(store.addRecording(
            filePath: saved.path, thumbnailSource: makePNGData(), cap: 50, maxAge: 30 * 24 * 60 * 60)) else { return }
        t.isTrue(store.savedFileExists(entry))
        try? FileManager.default.removeItem(at: saved)
        t.isFalse(store.savedFileExists(entry))
    },
]

File: windows/tests/BetterScreenshot.Tests/HistoryStoreTests.cs
using System.IO;
using System.Windows.Media;
using System.Windows.Media.Imaging;
using BetterScreenshot.History;
using BetterScreenshot.Platform;
using Xunit;

namespace BetterScreenshot.Tests;

public class HistoryStoreTests : IDisposable
{
    private readonly string _dir = Path.Combine(Path.GetTempPath(), "bs-hist-" + Guid.NewGuid().ToString("N"));

    public void Dispose()
    {
        try { if (Directory.Exists(_dir)) Directory.Delete(_dir, true); } catch { }
    }

    private static byte[] Png(int w = 60, int h = 40)
    {
        int stride = w * 4;
        var px = new byte[h * stride];
        for (int i = 0; i < px.Length; i += 4) { px[i] = 0; px[i + 1] = 0; px[i + 2] = 200; px[i + 3] = 255; }
        return ImageIo.EncodePng(BitmapSource.Create(w, h, 96, 96, PixelFormats.Bgra32, null, px, stride));
    }

    private string DummyRecording()
    {
        var path = Path.Combine(_dir, "rec-" + Guid.NewGuid().ToString("N") + ".mp4");
        Directory.CreateDirectory(_dir);
        File.WriteAllBytes(path, new byte[] { 0, 1, 2, 3 });
        return path;
    }

    [Fact]
    public void AddScreenshotWritesCopyThumbAndIndex()
    {
        var store = new HistoryStore(_dir, cap: 50);
        var entry = store.AddScreenshot(Png(), 50);
        Assert.NotNull(entry);
        Assert.True(File.Exists(store.ImagePath(entry!)!));
        Assert.True(File.Exists(store.ThumbPath(entry)));
        Assert.True(File.Exists(Path.Combine(_dir, "history.json")));
    }

    [Fact]
    public void AddRecordingStoresReferenceNotCopy()
    {
        var store = new HistoryStore(_dir, 50);
        var rec = DummyRecording();
        var entry = store.AddRecording(rec, Png(), 50);
        Assert.NotNull(entry);
        Assert.Equal(rec, entry!.FilePath);
        Assert.True(File.Exists(store.ThumbPath(entry)));
        Assert.True(File.Exists(rec)); // original untouched
        Assert.Null(entry.ImageFile);
    }

    [Fact]
    public void ReloadRoundTripsIndex()
    {
        var store = new HistoryStore(_dir, 50);
        var a = store.AddScreenshot(Png(), 50)!;
        var b = store.AddScreenshot(Png(), 50)!;
        var reloaded = new HistoryStore(_dir, 50);
        Assert.Equal(2, reloaded.Index.Entries.Count);
        Assert.Equal(b.Id, reloaded.Index.Entries[0].Id); // newest first
        Assert.Equal(a.Id, reloaded.Index.Entries[1].Id);
    }

    [Fact]
    public void CapEvictionDeletesOwnedFiles()
    {
        var store = new HistoryStore(_dir, cap: 1);
        var first = store.AddScreenshot(Png(), 1)!;
        var firstImage = store.ImagePath(first)!;
        var firstThumb = store.ThumbPath(first);
        store.AddScreenshot(Png(), 1);
        Assert.Single(store.Index.Entries);
        Assert.False(File.Exists(firstImage));
        Assert.False(File.Exists(firstThumb));
    }

    [Fact]
    public void RemoveNeverDeletesSavedRecordingFile()
    {
        var store = new HistoryStore(_dir, 50);
        var rec = DummyRecording();
        var entry = store.AddRecording(rec, Png(), 50)!;
        var thumb = store.ThumbPath(entry);
        store.Remove(entry.Id);
        Assert.False(File.Exists(thumb));
        Assert.True(File.Exists(rec));
    }

    [Fact]
    public void ClearAllEmptiesIndexAndDeletesOwnedFiles()
    {
        var store = new HistoryStore(_dir, 50);
        var a = store.AddScreenshot(Png(), 50)!;
        store.AddScreenshot(Png(), 50);
        store.ClearAll();
        Assert.Empty(store.Index.Entries);
        Assert.False(File.Exists(store.ImagePath(a)!));
        Assert.Empty(new HistoryStore(_dir, 50).Index.Entries);
    }

    [Fact]
    public void CorruptIndexStartsEmpty()
    {
        Directory.CreateDirectory(_dir);
        File.WriteAllText(Path.Combine(_dir, "history.json"), "{ not valid json ]");
        Assert.Empty(new HistoryStore(_dir, 50).Index.Entries);
    }

    [Fact]
    public void AgePruneAppliesAtLoad()
    {
        var old = new DateTime(2026, 1, 1, 0, 0, 0, DateTimeKind.Utc);
        var store = new HistoryStore(_dir, 50, now: old);
        store.AddScreenshot(Png(), 50, date: old, now: old);
        var reloaded = new HistoryStore(_dir, 50, now: old.AddDays(45));
        Assert.Empty(reloaded.Index.Entries);
    }

    [Fact]
    public void MissingRecordingFilePrunedAtLoad()
    {
        var store = new HistoryStore(_dir, 50);
        var rec = DummyRecording();
        store.AddRecording(rec, Png(), 50);
        File.Delete(rec);
        Assert.Empty(new HistoryStore(_dir, 50).Index.Entries);
    }

    [Fact]
    public void SavedFileExistsReflectsDisk()
    {
        var store = new HistoryStore(_dir, 50);
        var rec = DummyRecording();
        var entry = store.AddRecording(rec, Png(), 50)!;
        Assert.True(store.SavedFileExists(entry));
        File.Delete(rec);
        Assert.False(store.SavedFileExists(entry));
    }
}
